### PR TITLE
Fix getDeviceAndModuleOnBehalfOf to check if target device is in scope

### PR DESCRIPTION
**Pull request by varunpuranik**
Issue: EdgeHub allows child EdgeHubs to get device/module information for the devices in their scope (l3/$edgeHub can ask for l3child information to l4/$edgeHub). But a child EdgeHub should not be able to ask for information on identities that are not its descendants. But the current code did not do this check. Consequently a child edgeHub (say l3/$edgeHub) can ask for information on an identity that is not its child (say l4child) and be able to receive it. 

Fix: Before returning the information to the requesting child, also check if the target device/module is one of its descendants.

### CODE REVIEW

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/NestedDeviceScopeApiClient.cs [hunk 1]
```diff
@@ -165,10 +165,6 @@ async Task<ScopeResult> GetIdentityOnBehalfOfInternalAsync(Uri uri, string devic
                 Option<string> maybeAuthChain = await this.serviceIdentityHierarchy.GetAuthChain(onBehalfOfDevice);
                 string authChain = maybeAuthChain.Expect(() => new InvalidOperationException($"No valid authentication chain for {onBehalfOfDevice}"));
 
-                // We must pass the origin Edge ID along upstream, so the root
-                // can know which Edge to act OnBehalfOf when calling IoT Hub
-                msg.Headers.Add(Constants.OriginEdgeHeaderKey, onBehalfOfDevice);
```
Comment (richma-ms) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/NestedDeviceScopeApiClient.cs:170]: >msg.Headers.Add(Constants.OriginEdgeHeaderKey, onBehalfOfDevice); [](start = 16, length = 65)

please have someone make sure that adding a newly created leaf at the bottom level still works.  This is covered by the nested E2E tests, but wouldn't hurt to have someone run it manually at least once.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs [hunk 2]
```diff
@@ -16,6 +16,7 @@ namespace Microsoft.Azure.Devices.Edge.Hub.Http.Controllers
     using Microsoft.Extensions.Logging;
     using Microsoft.Extensions.Primitives;
     using Newtonsoft.Json;
+    using Org.BouncyCastle.Asn1.Cmp;
```
Comment (richma-ms) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:19]: >using Org.BouncyCastle.Asn1.Cmp; [](start = 4, length = 32)

seems extraneous (prob VS autofill?)

**Comment by richma-ms** [comment]
>             HttpAuthResult authResult = await authenticator.AuthenticateAsync(actorDeviceId, Option.Some(actorModuleId), Option.Some(request.AuthChain), this.HttpContext);

hmm....I think this technically won't be necessary anymore after the change, as you're only relying on the authenticator to perform AuthN.

---
Refers to: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:93 in 44769f0. [](commit_id = 44769f0115c1f531a55807f34f1250e8e2a59832, deletion_comment = False)

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs [hunk 3]
```diff
@@ -139,19 +164,16 @@ async Task<EdgeHubScopeResult> HandleGetDeviceAndModuleOnBehalfOfAsync(string ac
...
-                originatingEdge = originHeader.First();
+                originatingEdgeDevice = actorDeviceId;
             }
 
             // We must always forward the call further upstream first,
             // as this is invoked for refreshing an identity on-demand,
             // and we don't know whether our cache is out-of-date.
             IDeviceScopeIdentitiesCache identitiesCache = edgeHub.GetDeviceScopeIdentitiesCache();
-            await identitiesCache.RefreshServiceIdentityOnBehalfOf(targetId, originatingEdge);
+            await identitiesCache.RefreshServiceIdentityOnBehalfOf(targetId, originatingEdgeDevice);
```
Comment (richma-ms) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:176]: >originatingEdgeDevice [](start = 77, length = 21)

I think this still needs a default (current deviceId) for the bottom level case

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs (hunk 2 above)
Comment (varunpuranik) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:19]: good catch.. fixed

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs (hunk 3 above)
Comment (varunpuranik) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:162]: Emm, I am not sure I follow.. what should be set to the default?

Comment (richma-ms) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:160]: RefreshServiceIdentityOnBehalfOf() will throw if the OnBehalfOfDevice is empty, which can happen here if the caller supplies an empty authchain.  We can either param check for the empty case and return a meaningful error, or default the OnBehalfOfDevice to the current Edge device ID like the previous code did.


---
In reply to: [605145686](https://github.com/Azure/iotedge/pull/4735#discussion_r605145686) [](ancestors = 605145686)

**Review by richma-ms — APPROVED** [review]

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs (hunk 3 above)
Comment (varunpuranik) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs:160]: yes, and I default to actorDeviceId if the authchain is empty - https://github.com/Azure/iotedge/pull/4735/files#diff-b372abe320493b2949d978c985c8e09cf07491f23268ac948f81ac3bc344a8b8R154

**Pull request merged**